File: test_uphone/tests/RotateWorldTest/RotateWorldTest.cpp
#include "RotateWorldTest.h"
#include "../testResource.h"
#include "CCLabel.h"
#include "CCDirector.h"
#include "CCMenu.h"
#include "CCIntervalAction.h"

//------------------------------------------------------------------
//
// TestLayer
//
//------------------------------------------------------------------
void TestLayer::onEnter()
{
	__super::onEnter();

	float x,y;
	
	CGSize size = CCDirector::getSharedDirector()->getWinSize();
	x = size.width;
	y = size.height;

	//NSArray *array = [UIFont familyNames];
	//for( NSString *s in array )
	//	NSLog( s );
	CCLabel* label = CCLabel::labelWithString("cocos2d", "Tahoma", 64);

	label->setPosition( CGPointMake(x/2,y/2) );
	
	addChild(label);
}

//------------------------------------------------------------------
//
// SpriteLayer
//
//------------------------------------------------------------------
void SpriteLayer::onEnter()
{
	__super::onEnter();

	float x,y;
	
	CGSize size = CCDirector::getSharedDirector()->getWinSize();
	x = size.width;
	y = size.height;
	
	CCSprite* sprite = CCSprite::spriteWithFile(s_pPathGrossini);
	CCSprite* spriteSister1 = CCSprite::spriteWithFile(s_pPathSister1);
	CCSprite* spriteSister2 = CCSprite::spriteWithFile(s_pPathSister2);
	
	sprite->setScale(1.5f);
	spriteSister1->setScale(1.5f);
	spriteSister2->setScale(1.5f);
	
	sprite->setPosition(CGPointMake(x/2,y/2));
	spriteSister1->setPosition(CGPointMake(40,y/2));
	spriteSister2->setPosition(CGPointMake(x-40,y/2));

	CCAction *rot = CCRotateBy::actionWithDuration(16, -3600);
	
	addChild(sprite);
	addChild(spriteSister1);
	addChild(spriteSister2);
	
	sprite->runAction(rot);

	CCIntervalAction *jump1 = CCJumpBy::actionWithDuration(4, CGPointMake(-400,0), 100, 4);
	CCIntervalAction *jump2 = jump1->reverse();
	
	CCIntervalAction *rot1 = CCRotateBy::actionWithDuration(4, 360*2);
	CCIntervalAction *rot2 = rot1->reverse();
	
	spriteSister1->runAction(CCRepeat::actionWithAction( CCSequence::actions(jump2, jump1, NULL), 5 ));
	spriteSister2->runAction(CCRepeat::actionWithAction( CCSequence::actions(dynamic_cast<CCFiniteTimeAction *>(jump1->copy()->autorelease()), dynamic_cast<CCFiniteTimeAction *>(jump2->copy()->autorelease()), NULL), 5 ));
	
	spriteSister1->runAction(CCRepeat::actionWithAction( CCSequence::actions(rot1, rot2, NULL), 5 ));
	spriteSister2->runAction(CCRepeat::actionWithAction( CCSequence::actions(dynamic_cast<CCFiniteTimeAction *>(rot2->copy()->autorelease()), dynamic_cast<CCFiniteTimeAction *>(rot1->copy()->autorelease()), NULL), 5 ));
}

//------------------------------------------------------------------
//
// RotateWorldMainLayer
//
//------------------------------------------------------------------

void RotateWorldMainLayer::onEnter()
{
	__super::onEnter();

	float x,y;
	
	CGSize size = CCDirector::getSharedDirector()->getWinSize();
	x = size.width;
	y = size.height;
	
	CCNode* blue =  CCColorLayer::layerWithColor(ccc4(0,0,255,255));
	CCNode* red =   CCColorLayer::layerWithColor(ccc4(255,0,0,255));
	CCNode* green = CCColorLayer::layerWithColor(ccc4(0,255,0,255));
	CCNode* white = CCColorLayer::layerWithColor(ccc4(255,255,255,255));

	blue->setScale(0.5f);
	blue->setPosition(CGPointMake(-x/4,-y/4));
	blue->addChild( SpriteLayer::node() );
	
	red->setScale(0.5f);
	red->setPosition(CGPointMake(x/4,-y/4));

	green->setScale(0.5f);
	green->setPosition(CGPointMake(-x/4,y/4));
	green->addChild(TestLayer::node());

	white->setScale(0.5f);
	white->setPosition(CGPointMake(x/4,y/4));

	addChild(blue, -1);
	addChild(white);
	addChild(green);
	addChild(red);

	CCAction* rot = CCRotateBy::actionWithDuration(8, 720);
	
	blue->runAction(rot);
	red->runAction( dynamic_cast<CCAction *>(rot->copy()->autorelease()) );
	green->runAction( dynamic_cast<CCAction *>(rot->copy()->autorelease()) );
	white->runAction( dynamic_cast<CCAction *>(rot->copy()->autorelease()) );
}

void RotateWorldTestScene::runThisTest()
{
    CCLayer* pLayer = RotateWorldMainLayer::node();

    addChild(pLayer);
    runAction( CCRotateBy::actionWithDuration(4, -360) );

    CCDirector::getSharedDirector()->replaceScene(this);

}

File: samples/lua-tests/src/RotateWorldTest/RotateWorldTest.lua

local size = cc.Director:getInstance():getWinSize()

local function CreateSpriteLayer()
	local layer = cc.Layer:create()

	local x, y
    x = size.width
    y = size.height

    local sprite = cc.Sprite:create(s_pPathGrossini)
    local spriteSister1 = cc.Sprite:create(s_pPathSister1)
    local spriteSister2 = cc.Sprite:create(s_pPathSister2)

    sprite:setScale(1.5)
    spriteSister1:setScale(1.5)
    spriteSister2:setScale(1.5)

    sprite:setPosition(cc.p(x / 2, y / 2))
    spriteSister1:setPosition(cc.p(40, y / 2))
    spriteSister2:setPosition(cc.p(x - 40, y / 2))

    layer:addChild(sprite)
    layer:addChild(spriteSister1)
    layer:addChild(spriteSister2)

	local rot = cc.RotateBy:create(16, -3600)
    sprite:runAction(rot)

    local jump1 = cc.JumpBy:create(4, cc.p(-400, 0), 100, 4)
    local jump2 = jump1:reverse()

    local rot1 = cc.RotateBy:create(4, 360 * 2)
    local rot2 = rot1:reverse()

    local jump3 = cc.JumpBy:create(4, cc.p(-400, 0), 100, 4)
    local jump4 = jump3:reverse()
    local rot3 = cc.RotateBy:create(4, 360 * 2)
    local rot4 = rot3:reverse()

    spriteSister1:runAction(cc.Repeat:create(cc.Sequence:create(jump2, jump1), 5))
    spriteSister2:runAction(cc.Repeat:create(cc.Sequence:create(jump3, jump4), 5))

    spriteSister1:runAction(cc.Repeat:create(cc.Sequence:create(rot1, rot2), 5))
    spriteSister2:runAction(cc.Repeat:create(cc.Sequence:create(rot4, rot3), 5))

	return layer
end

local function CreateTestLayer()
	local layer = cc.Layer:create()

	local x, y
    x = size.width
    y = size.height

    local label = cc.LabelTTF:create("cocos2d", "Tahoma", 64)
    label:setPosition(x / 2, y / 2)
    layer:addChild(label)

	return layer
end

local function CreateRotateWorldLayer()
	local layer = cc.Layer:create()

	local x, y
    x = size.width
    y = size.height

    local blue =  cc.LayerColor:create(cc.c4b(0,0,255,255))
    local red =   cc.LayerColor:create(cc.c4b(255,0,0,255))
    local green = cc.LayerColor:create(cc.c4b(0,255,0,255))
    local white = cc.LayerColor:create(cc.c4b(255,255,255,255))

    blue:setScale(0.5)
    blue:setPosition(cc.p(- x / 4, - y / 4))
    blue:addChild(CreateSpriteLayer())

    red:setScale(0.5)
    red:setPosition(cc.p(x / 4, - y / 4))

    green:setScale(0.5)
    green:setPosition(cc.p(- x / 4, y / 4))
    green:addChild(CreateTestLayer())

    white:setScale(0.5)
    white:setPosition(cc.p(x / 4, y / 4))
    white:ignoreAnchorPointForPosition(false)
    white:setPosition(cc.p(x / 4 * 3, y / 4 * 3))

    layer:addChild(blue, -1)
    layer:addChild(white)
    layer:addChild(green)
    layer:addChild(red)

    local rot = cc.RotateBy:create(8, 720)
    local rot1 = cc.RotateBy:create(8, 720)
    local rot2 = cc.RotateBy:create(8, 720)
    local rot3 = cc.RotateBy:create(8, 720)

    blue:runAction(rot)
    red:runAction(rot1)
    green:runAction(rot2)
    white:runAction(rot3)

	return layer
end

--------------------------------
-- Rotate World Test
--------------------------------
function RotateWorldTest()
	cclog("RotateWorldTest")
	local scene = cc.Scene:create()
	local layer = CreateRotateWorldLayer()

	scene:addChild(layer)
	scene:addChild(CreateBackMenuItem())
	scene:runAction(cc.RotateBy:create(4, -360))

	return scene
end
